### DISCUSSION TITLE
简单介绍 Leaf 的请求分发机制

### DISCUSSION THREAD

**Discussion by eycorsican** [eycorsican/leaf#41] [Show and tell]
https://tachyondevel.medium.com/%E7%AE%80%E5%8D%95%E4%BB%8B%E7%BB%8D-leaf-%E7%9A%84%E8%AF%B7%E6%B1%82%E5%88%86%E5%8F%91%E6%9C%BA%E5%88%B6-cecde244afb

**Comment by ghost** [comment]
老哥 文章写得很透彻

**Comment by darling3m** [comment]
能不能写个教程怎么配置

**Comment by LazyZhu** [comment]
Leaf 的请求分发机制目前看来是比较优秀的, 但就像文中提出的一样 Leaf 经过一次性的测量方式选出的不一定是最优的, 这在实际使用是就毫无意义了, 就像从网上买苹果, 从产地,价格等方面货比三家, 但送来的却是烂苹果. 这不是 Leaf 的问题, 网络状况如此, 所以希望 Leaf 像 ss-rust 一样改进测量方式:
https://github.com/shadowsocks/shadowsocks-rust/blob/master/crates/shadowsocks-service/src/local/loadbalancing/server_stat.rs#L60-L81

>     fn score(&self) -> u32 {
>         // Normalize rtt
>         let nrtt = self.rtt as f64 / MAX_SERVER_RTT as f64;
> 
>         // Normalize stdev
>         let nstdev = self.latency_stdev / max_latency_stdev();
> 
>         const SCORE_RTT_WEIGHT: f64 = 1.0;
>         const SCORE_FAIL_WEIGHT: f64 = 3.0;
>         const SCORE_STDEV_WEIGHT: f64 = 1.0;
> 
>         // Score = (norm_lat * 1.0 + prop_err * 3.0 + stdev * 1.0) / 5.0
>         //
>         // 1. The lower latency, the better
>         // 2. The lower errored count, the better
>         // 3. The lower latency's stdev, the better
>         let score = (nrtt * SCORE_RTT_WEIGHT + self.fail_rate * SCORE_FAIL_WEIGHT + nstdev * SCORE_STDEV_WEIGHT)
>             / (SCORE_RTT_WEIGHT + SCORE_FAIL_WEIGHT + SCORE_STDEV_WEIGHT);
> 
>         // Times 10000 converts to u32, for 0.0001 precision
>         (score * 10000.0) as u32
>     }
>

**Reply by eycorsican** [reply]
欢迎来加入其它测量方法啊，加一个 `checkAlgorithm` 参数切换不同测量方法即可。

我个人并不太在意这个，测个大概就行，因为第一个不行还会尝试第二个、第三个出口。leaf 这方面提供的灵活性甚至不需要健康检测也能够满足很多使用场景，我目前使用的配置就没用到健康检测（health-check=false）：
```ini
kt1_tro = retry, p_kt1_tro, attempts=2
lb = failover, kt1_tro, SS, kt1_tro, health-check=false, fail-timeout=3, failover=true, fallback-cache=true, cache-timeout=3
lb2 = failover, SS, kt1_tro, health-check=false, fail-timeout=2, failover=true
```

**Reply by LazyZhu** [reply]
个人很有这个意愿添加不同测量方法, 但是一则刚接触 Rust, 而 Rust 又不像 Golang 那么简明, 再则本人本职不是程序员, 编程只是个人爱好而已, 过年有时间恶补下 Rust 后看情况吧 :)
